tools/generator/dfg/sam/sam_identifier.py

```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2013     , Kevin Läufer
+# Copyright (c) 2013-2014, Niklas Hauser
+# Copyright (c)      2016, Fabian Greif
+# All rights reserved.
+
+import re
+import logging
+
+from modm_devices.device_identifier import DeviceIdentifier
+
+LOGGER = logging.getLogger("dfg.sam.identifier")
+
+class SAMIdentifier:
+    """ SAMIdentifier
+    A class to parse SAM device strings, e.g. ATSAMD21E15A-MUT.
+    Device names are organized as follows:
+       ATSAMD     21         E        15        A     -    M          U           T
+      {family} {Series} {Pin Count} {Flash} {Revision} {pacakge} {Temp Grade} {carrier}
+    """
```

Comment:
For reference, this is how the datasheet splits it up:
![58F9B61D-3D0D-4A77-B549-165122E49CFF](https://user-images.githubusercontent.com/163066/56044161-e00fd400-5d3e-11e9-8123-7ce194db7fad.jpeg)

I’d split it a little differently:
```
    sam      d        21    e     15      a     -    m
{platform}{family}{series}{pin}{flash}{revision}-{package}
```
You also need the `{package}`, since they bond the same silicon die differently for the packages, and then some pins may not be available in a different package.

The temperature grade and carrier likely don't have an impact on the device structure, so I'd leave them out for now (we can simply add them later if they turn out important).

Reply:
That looks like the old Atmel Datasheet, the new one is [here](http://ww1.microchip.com/downloads/en/DeviceDoc/SAMD21-Family-DataSheet-DS40001882D.pdf)

As far as I can tell the package does not make a difference, with one exception, the WLCSP package. The `.atdf` files reflect this since they either have no package type in the file name or there is a `U` appended to the end.

![image](https://user-images.githubusercontent.com/9755578/56057187-159ec800-5d13-11e9-8a64-222bfa972425.png)

Reply:
So then it would just be like this?
```
   atsam     d        21    e     15      a
{platform}{family}{series}{pin}{flash}{revision}
```

Reply:
I was assuming that the name was coming from the `atdf` filename, which I think was a bad assumption. the XML _inside_ the `atdf` file contains full part numbers. so:
```
    sam         d      21        e    15         a    -    m
{platform}{family}{series}{pin}{flash}{revision}{package}
```
then group according to U and NON-U in `sam_groups.py`

Also, would you prefer for the capture names to match the datasheet or the other python files?
example: change series->name, flash->size like in the stm32 parser

Reply:
No, namings should be as close as possible/reasonable to the datasheets. Apart from `{platform}` there is no cross-vendor naming scheme. It's then up to the HAL implementation to rename/reformat the names into a consistent form.